dds/DCPS/transport/framework/DataLink.h

```diff
@@ -250,7 +250,7 @@ class OpenDDS_Dcps_Export DataLink
   /// targets of this DataLink (see is_target()).
   GUIDSeq* target_intersection(const RepoId& pub_id, const GUIDSeq& in, size_t& n_subs);
 
-  TransportImpl& impl() const;
+  RcHandle<TransportImpl> impl() const;
```

Comment:
Return the `WeakRcHandle`.  This will force the using code to come to grips with the reality that the transport might not be there.

Reply:
I think this would cause undue pain... we should be checking the return value, but having callers perform their own lock / check seems like it would generate a lot of clutter...

Reply:
The clutter already has to be there, i.e., every caller of `impl` has to check that it is null.  By returning a WeakRcHandle, the compiler ensures that every caller is performing the check.

Reply:
Both the strong and weak handles should be checked by the caller.  I agree there's some advantage to using the weak handle as the return type in order to force callers to do this.  However, it is different in terms of semantics of how the objects relate to each other.  Here's an example:

```
class A {
public:
  SomeHandle<B> op();
private:
  WeakRcHandle<B> b_;
  // other members, mutex
};
```

In this scenario, we know that A's data member has to be a weak handle.  Within the method `op()`, it acquires the mutex and examines other internal state.  As part of this method it locks the weak handle.  If this fails, the method must return a null handle (of either kind).  If it succeeds, returning the strong handle means that extending the lifetime of B was one of the postconditions of this method.  Returning the weak handle gives the caller the potential to use B, but if no other strong pointer to B exists then the object goes away before the caller can do its own locking.

Reply:
I've added checks for most uses now, but the config() calls are a bit trickier because the config signature (returns a reference) assumes it can't fail. This probably needs to become a RCH as well...

Reply:
OK, configs are RCH now as well, and should be checked most of the time before use now.